File: src/test/java/unittests/ConsTest.java
package unittests;

import core.procedures.cons.Car;
import core.procedures.cons.Cdr;
import org.junit.Test;

import java.util.Collections;

import static core.procedures.cons.ConsProc.Companion;
import static core.procedures.predicates.Predicate.*;
import static core.scm.Cons.EMPTY;
import static core.scm.Cons.list;
import static java.lang.Boolean.FALSE;
import static java.lang.Boolean.TRUE;
import static org.junit.Assert.*;

public class ConsTest {

  @Test
  public void testEquality() {
    assertTrue(EMPTY == Companion.cons(null, null));
    assertTrue(Companion.cons(null, null) == Companion.cons(null, null));
    assertEquals(EMPTY, Companion.cons(null, null));
    assertEquals(Companion.cons(null, null), Companion.cons(null, null));
    assertEquals(Companion.cons(1, 2), Companion.cons(1, 2));
    assertFalse(Companion.cons(1, 2) == Companion.cons(1, 2));
  }

  @Test
  public void testToString() {
    assertEquals("()", EMPTY.toString());
    assertEquals("()", Companion.cons(null, null).toString());
    assertEquals("(1 . 2)", Companion.cons(1, 2).toString());
    assertEquals("(1 2 . 3)", Companion.cons(1, Companion.cons(2, 3)).toString());
    assertEquals("(1 2 3 . 4)", Companion.cons(1, Companion.cons(2, (Companion.cons(3, 4)))).toString());
    assertEquals("(1)", Companion.cons(1, EMPTY).toString());
    assertEquals("(1 2)", Companion.cons(1, Companion.cons(2, EMPTY)).toString());
    assertEquals("(1 2 3)", Companion.cons(1, Companion.cons(2, Companion.cons(3, EMPTY))).toString());
    assertEquals("(1 2 3 4)", Companion.cons(1, Companion.cons(2, Companion.cons(3, Companion.cons(4, EMPTY)))).toString());
    assertEquals("(())", Companion.cons(EMPTY, EMPTY).toString());
    assertEquals("(() ())", Companion.cons(EMPTY, Companion.cons(EMPTY, EMPTY)).toString());
    assertEquals("(() . 1)", Companion.cons(EMPTY, 1).toString());
    assertEquals("(() (1 2 ()) ())", list(EMPTY, list(1, 2, EMPTY), EMPTY).toString());
    assertEquals("()", list().toString());
    assertEquals("(1)", list(1).toString());
    assertEquals("(1)", Companion.cons(1, Collections.EMPTY_LIST).toString());
    assertEquals("(2 1)", Companion.cons(2, Companion.cons(1, Collections.EMPTY_LIST)).toString());
    assertEquals("(1 2)", list(1, 2).toString());
    assertEquals("(1 2 3 4)", list(1, 2, 3, 4).toString());
    assertEquals("(1 2 () 4)", list(1, 2, EMPTY, 4).toString());
  }

  @Test
  public void testList() {
    assertEquals(Companion.cons(1, Companion.cons(2, Companion.cons(3, EMPTY))), list(1, 2, 3));
    assertEquals(list(1, 2, 3) ,list(1, 2, 3));
    assertEquals(Companion.cons(1, EMPTY), list(1));
    assertEquals(EMPTY, list());
    assertEquals(Companion.cons(EMPTY, Companion.cons(EMPTY, Companion.cons(EMPTY, EMPTY))), list(EMPTY, EMPTY, EMPTY));
  }

  @Test
  public void testCar() {
    assertEquals(1, Car.Companion.car(Companion.cons(1, 2)));
    assertEquals(1, Car.Companion.car(Companion.cons(1, Companion.cons(2, 3))));
    assertEquals(EMPTY, Car.Companion.car(Companion.cons(EMPTY, Companion.cons(2, 3))));

    assertEquals(1, Car.Companion.car(list(1)));
    assertEquals(1, Car.Companion.car(list(1, 2)));
    assertEquals(1, Car.Companion.car(list(1, 2, 3)));

    assertEquals(2, Car.Companion.car(Cdr.Companion.cdr(list(1, 2, 3))));
    assertEquals(3, Car.Companion.car(Cdr.Companion.cdr(Cdr.Companion.cdr(list(1, 2, 3)))));

    try {
      Car.Companion.car(EMPTY);
      fail();
    } catch (IllegalArgumentException e) {
      assertEquals("car: type mismatch; (expected: Pair, given: ())", e.getMessage());
    }
    try {
      Car.Companion.car(list());
      fail();
    } catch (IllegalArgumentException e) {
      assertEquals("car: type mismatch; (expected: Pair, given: ())", e.getMessage());
    }
  }

  @Test
  public void testCdr() {
    assertEquals(2, Car.Companion.car(Cdr.Companion.cdr(Companion.cons(1, Companion.cons(2, 3)))));
    assertEquals(3, Cdr.Companion.cdr(Cdr.Companion.cdr(Companion.cons(1, Companion.cons(2, 3)))));
    assertEquals(3, Cdr.Companion.cdr(Cdr.Companion.cdr(Companion.cons(1, Companion.cons(2, 3)))));
    assertEquals(EMPTY, Cdr.Companion.cdr(list(1)));
    assertEquals(EMPTY, Cdr.Companion.cdr(list((Object)EMPTY)));
    assertEquals(list(2, 3), Cdr.Companion.cdr(list(1, 2, 3)));
    assertEquals(list(3), Cdr.Companion.cdr(Cdr.Companion.cdr(list(1, 2, 3))));
    assertEquals(EMPTY, Cdr.Companion.cdr(Cdr.Companion.cdr(Cdr.Companion.cdr(list(1, 2, 3)))));
  }

  @Test
  public void testLength() {
    assertEquals(0, list().size());
    assertEquals(1, list(1).size());
    assertEquals(2, list(EMPTY, Companion.cons(EMPTY, EMPTY)).size());
    assertEquals(3, list(1, EMPTY, 3).size());
    assertEquals(9, list(1, EMPTY, 3, 4, 5, 6, 777, 88, 99999).size());
  }

  @Test
  public void testIsNil() {
    assertEquals(TRUE,  IS_NULL.apply((Object)null));
    assertEquals(FALSE, IS_NULL.apply(EMPTY));
    assertEquals(FALSE, IS_NULL.apply(Companion.cons(1, null)));
    assertEquals(FALSE, IS_NULL.apply(Companion.cons(null, 2)));
    assertEquals(FALSE, IS_NULL.apply(Companion.cons(EMPTY, 2)));
    assertEquals(FALSE, IS_NULL.apply(Companion.cons(EMPTY, EMPTY)));
    assertEquals(FALSE, IS_NULL.apply(Companion.cons(1, EMPTY)));
    assertEquals(FALSE, IS_NULL.apply(Companion.cons(1, Companion.cons(2, 3))));
    assertEquals(FALSE, IS_NULL.apply(list((Object) null)));
    assertEquals(FALSE, IS_NULL.apply(list((Object)EMPTY)));
    assertEquals(FALSE, IS_NULL.apply(list(1)));
    assertEquals(FALSE, IS_NULL.apply(list(1, 2)));
    assertEquals(FALSE, IS_NULL.apply(list(1, 2, 3)));
    assertEquals(FALSE, IS_NULL.apply(1));
    assertEquals(FALSE, IS_NULL.apply("test"));
  }

  @Test
  public void testIsPair() {
    assertEquals(FALSE, IS_PAIR.apply(EMPTY));
    assertEquals(FALSE, IS_PAIR.apply(Companion.cons(null, null)));
    assertEquals(FALSE, IS_PAIR.apply(1));
    assertEquals(FALSE, IS_PAIR.apply("test"));
    assertEquals(TRUE, IS_PAIR.apply(Companion.cons(1, 2)));
    assertEquals(TRUE, IS_PAIR.apply(Companion.cons(1, EMPTY)));
    assertEquals(TRUE, IS_PAIR.apply(Companion.cons(EMPTY, EMPTY)));
    assertEquals(TRUE, IS_PAIR.apply(Companion.cons(1, Companion.cons(2, 3))));
    assertEquals(TRUE, IS_PAIR.apply(Companion.cons(1, Companion.cons(2, Companion.cons(3, 4)))));
    assertEquals(FALSE, IS_PAIR.apply(list()));
    assertEquals(TRUE, IS_PAIR.apply(list(1)));
    assertEquals(TRUE, IS_PAIR.apply(list(1, 2)));
    assertEquals(TRUE, IS_PAIR.apply(list(1, 2, 3)));
  }

  @Test
  public void testIsList() {
    assertEquals(TRUE,  IS_LIST.apply(EMPTY));
    assertEquals(TRUE,  IS_LIST.apply(list()));
    assertEquals(TRUE,  IS_LIST.apply(list(1)));
    assertEquals(TRUE,  IS_LIST.apply(list(1, 2)));
    assertEquals(TRUE,  IS_LIST.apply(list(1, 2, 3)));
    assertEquals(TRUE,  IS_LIST.apply(list(1, 2, 3, 4)));
    assertEquals(TRUE,  IS_LIST.apply(list(1, 2, EMPTY)));
    assertEquals(TRUE,  IS_LIST.apply(Companion.cons(null, null)));
    assertEquals(TRUE,  IS_LIST.apply(Companion.cons(1, null)));
    assertEquals(FALSE, IS_LIST.apply(Companion.cons(null, 1)));
    assertEquals(FALSE, IS_LIST.apply(Companion.cons(1, 2)));
    assertEquals(FALSE, IS_LIST.apply(Companion.cons(1, Companion.cons(2, Companion.cons(3, 4)))));
    assertEquals(TRUE,  IS_LIST.apply(Companion.cons(1, Companion.cons(2, Companion.cons(3, EMPTY)))));
    assertEquals(TRUE,  IS_LIST.apply(Companion.cons(1, Companion.cons(2, Companion.cons(3, list(1, 2, 3))))));
  }
}

File: src/test/java/unittests/ConsTest.kt
package unittests

import core.procedures.cons.Car
import core.procedures.cons.Cdr
import core.procedures.cons.ConsProc
import core.procedures.predicates.Predicate
import core.writer.Writer
import org.junit.Assert.*
import org.junit.Test

class ConsTest {

    private val car  = Car()
    private val cdr  = Cdr()
    private val cons = ConsProc()

    @Test
    fun testEquality() {
        assertEquals(cons(null, null), cons(null, null))
        assertEquals(cons(1, 2), cons(1, 2))
        assertFalse(cons(1, 2) === cons(1, 2))
    }

    @Test
    fun testToString() {
        assertEquals("(1 . 2)", Writer.write(cons(1, 2)))
        assertEquals("(1 2 . 3)", Writer.write(cons(1, cons(2, 3))))
        assertEquals("(1 2 3 . 4)", Writer.write(cons(1, cons(2, cons(3, 4)))))
        assertEquals("(1)",       Writer.write(cons(1, emptyList<Nothing>())))
        assertEquals("(1 2)",     Writer.write(cons(1, cons(2, emptyList<Nothing>()))))
        assertEquals("(1 2 3)",   Writer.write(cons(1, cons(2, cons(3, emptyList<Nothing>())))))
        assertEquals("(1 2 3 4)", Writer.write(cons(1, cons(2, cons(3, cons(4, emptyList<Nothing>()))))))
        assertEquals("(())",      Writer.write(cons(emptyList<Nothing>(), emptyList<Nothing>())))
        assertEquals("(() ())",   Writer.write(cons(emptyList<Nothing>(), cons(emptyList<Nothing>(), emptyList<Nothing>()))))
        assertEquals("(() . 1)",  Writer.write(cons(emptyList<Nothing>(), 1)))
        assertEquals("(() (1 2 ()) ())", Writer.write(listOf(emptyList<Nothing>(), listOf(1, 2, emptyList<Nothing>()), emptyList<Nothing>())))
        assertEquals("()",    Writer.write(listOf<Any>()))
        assertEquals("(1)",   Writer.write(listOf(1)))
        assertEquals("(1)",   Writer.write(cons(1, emptyList<Any?>())))
        assertEquals("(2 1)", Writer.write(cons(2, cons(1, emptyList<Any?>()))))
        assertEquals("(1 2)", Writer.write(listOf(1, 2)))
        assertEquals("(1 2 3 4)",  Writer.write(listOf(1, 2, 3, 4)))
        assertEquals("(1 2 () 4)", Writer.write(listOf(1, 2, emptyList<Nothing>(), 4)))
    }

    @Test
    fun testList() {
        assertEquals(cons(1, cons(2, cons(3, emptyList<Nothing>()))), listOf(1, 2, 3))
        assertEquals(listOf(1, 2, 3), listOf(1, 2, 3))
        assertEquals(cons(1, emptyList<Nothing>()), listOf(1))
        assertEquals(emptyList<Nothing>(), listOf<Any>())
        assertEquals(cons(emptyList<Nothing>(), cons(emptyList<Nothing>(), cons(emptyList<Nothing>(), emptyList<Nothing>()))), listOf(emptyList<Nothing>(), emptyList<Nothing>(), emptyList<Nothing>()))
    }

    @Test
    fun testCar() {
        assertEquals(1, car(cons(1, 2)))
        assertEquals(1, car(cons(1, cons(2, 3))))
        assertEquals(emptyList<Nothing>(), car(cons(emptyList<Nothing>(), cons(2, 3))))

        assertEquals(1, car(listOf(1)))
        assertEquals(1, car(listOf(1, 2)))
        assertEquals(1, car(listOf(1, 2, 3)))

        assertEquals(2, car(cdr(listOf(1, 2, 3))))
        assertEquals(3, car(cdr(cdr(listOf(1, 2, 3)))))
    }

    @Test
    fun testCdr() {
        assertEquals(2, car(cdr(cons(1, cons(2, 3)))))
        assertEquals(3, cdr(cdr(cons(1, cons(2, 3)))))
        assertEquals(3, cdr(cdr(cons(1, cons(2, 3)))))
        assertEquals(emptyList<Nothing>(), cdr(listOf(1)))
        assertEquals(emptyList<Nothing>(), cdr(listOf(emptyList<Nothing>() as Any)))
        assertEquals(listOf(2, 3), cdr(listOf(1, 2, 3)))
        assertEquals(listOf(3), cdr(cdr(listOf(1, 2, 3))))
        assertEquals(emptyList<Nothing>(), cdr(cdr(cdr(listOf(1, 2, 3)))))
    }

    @Test
    fun testLength() {
        assertEquals(0, listOf<Any>().size.toLong())
        assertEquals(1, listOf(1).size.toLong())
        assertEquals(2, listOf<List<*>>(emptyList<Nothing>(), cons(emptyList<Nothing>(), emptyList<Nothing>())).size.toLong())
        assertEquals(3, listOf(1, emptyList<Nothing>(), 3).size.toLong())
    }

    @Test
    fun testIsNil() {
        assertEquals(true,  Predicate.IS_NULL(null))
        assertEquals(false, Predicate.IS_NULL(emptyList<Nothing>()))
        assertEquals(false, Predicate.IS_NULL(cons(1, null)))
        assertEquals(false, Predicate.IS_NULL(cons(emptyList<Nothing>(), 2)))
        assertEquals(false, Predicate.IS_NULL(cons(emptyList<Nothing>(), emptyList<Nothing>())))
        assertEquals(false, Predicate.IS_NULL(cons(1, emptyList<Nothing>())))
        assertEquals(false, Predicate.IS_NULL(cons(1, cons(2, 3))))
        assertEquals(false, Predicate.IS_NULL(listOf<Any?>(null)))
        assertEquals(false, Predicate.IS_NULL(listOf(emptyList<Nothing>())))
        assertEquals(false, Predicate.IS_NULL(listOf(1)))
        assertEquals(false, Predicate.IS_NULL(listOf(1, 2)))
        assertEquals(false, Predicate.IS_NULL(listOf(1, 2, 3)))
        assertEquals(false, Predicate.IS_NULL(1))
        assertEquals(false, Predicate.IS_NULL("test"))
    }

    @Test
    fun testIsPair() {
        assertEquals(false, Predicate.IS_PAIR(emptyList<Nothing>()))
        assertEquals(false, Predicate.IS_PAIR(1))
        assertEquals(false, Predicate.IS_PAIR("test"))
        assertEquals(false, Predicate.IS_PAIR(listOf<Any>()))
        assertEquals(true,  Predicate.IS_PAIR(cons(null, null)))
        assertEquals(true,  Predicate.IS_PAIR(cons(1, 2)))
        assertEquals(true,  Predicate.IS_PAIR(cons(1, emptyList<Nothing>())))
        assertEquals(true,  Predicate.IS_PAIR(cons(emptyList<Nothing>(), emptyList<Nothing>())))
        assertEquals(true,  Predicate.IS_PAIR(cons(1, cons(2, 3))))
        assertEquals(true,  Predicate.IS_PAIR(cons(1, cons(2, cons(3, 4)))))
        assertEquals(true,  Predicate.IS_PAIR(listOf(1)))
        assertEquals(true,  Predicate.IS_PAIR(listOf(1, 2)))
        assertEquals(true,  Predicate.IS_PAIR(listOf(1, 2, 3)))
    }

    @Test
    fun testIsList() {
        assertEquals(true,  Predicate.IS_LIST(emptyList<Nothing>()))
        assertEquals(true,  Predicate.IS_LIST(listOf<Any>()))
        assertEquals(true,  Predicate.IS_LIST(listOf(1)))
        assertEquals(true,  Predicate.IS_LIST(listOf(1, 2)))
        assertEquals(true,  Predicate.IS_LIST(listOf(1, 2, 3)))
        assertEquals(true,  Predicate.IS_LIST(listOf(1, 2, 3, 4)))
        assertEquals(true,  Predicate.IS_LIST(listOf(1, 2, emptyList<Nothing>())))
        assertEquals(true,  Predicate.IS_LIST(cons(null, null)))
        assertEquals(true,  Predicate.IS_LIST(cons(1, null)))
        assertEquals(false, Predicate.IS_LIST(cons(1, 2)))
        assertEquals(false, Predicate.IS_LIST(cons(1, cons(2, cons(3, 4)))))
        assertEquals(true,  Predicate.IS_LIST(cons(1, cons(2, cons(3, emptyList<Nothing>())))))
        assertEquals(true,  Predicate.IS_LIST(cons(1, cons(2, cons(3, listOf(1, 2, 3))))))
    }
}
